### DISCUSSION TITLE
websocket and session

### DISCUSSION THREAD

**Discussion by dineshbvadhia** [zauberzeug/nicegui#3901] [Q&A] [answered]
### Question

My understanding is a websocket facilitates bi-directional communication between a client and an endpoint on a server.

Imagine a client UI which maps to different server endpoints. Can the interaction between the user and the server to the different endpoints take place as a websocket session (assuming there is such a thing)?

**Comment by bitnom** [comment]
> ### Question
> My understanding is a websocket facilitates bi-directional communication between a client and an endpoint on a server.
> 
> Imagine a client UI which maps to different server endpoints. Can the interaction between the user and the server to the different endpoints take place as a websocket session (assuming there is such a thing)?

I'm new to nicegui, and Idk if it's using websockets (I hope it is). Generally though, the answer to your question is yes. You would use a JWT token authentication scheme. That way you can have multiple servers authenticate the user but they don't have to callback to a central authentication server. They just have to verify the authenticity of the token and claims within (After decrypting it with a private key)

**Comment by rodja** [comment]
> My understanding is a websocket facilitates bi-directional communication between a client and an endpoint on a server.

Yes, NiceGUI uses a websocket (via socketio) to do the bi-directional communication.

> Imagine a client UI which maps to different server endpoints. Can the interaction between the user and the server to the different endpoints take place as a websocket session (assuming there is such a thing)?

I'm not sure what the question is. If you have one ore more servers with RESTful HTTP APIs (like imdb, youtube, ...) and want to show these infos in a NiceGUI application you can simply write Python code to access these APIs. Updates to the UI are automatically synced to the browser via NiceGUIs websocket communication.

**Reply by dineshbvadhia** [reply]
@rodja From your reponse, the answer is yes but to clarify:

For example, a server has 3 endpoints:
```
/this
/that
/other
```
The browser UI has 3 buttons: `<this>, <that>, <other>`

A (registered or anonymous) user begins a browser session. They press the 3 buttons continuously for a few minutes before exiting the client app.

Typically, the browser-server session will consist of a sequence of http connections between the button and endpoint. The http connections can be replaced with websocket connections. The question is, can the browser-server session be a websocket session so all the communication takes place as a websocket and not http?

**Reply by rodja** [reply, marked as answer]
> For example, a server has 3 endpoints ... the browser UI has 3 buttons

So in your scenario the endpoints are implemented with FastAPI "inside" the NiceGUI app? Only one server?

> Typically, the browser-server session will consist of a sequence of http connections between the button and endpoint.

Not with NiceGUI. Here each button press is an event over a persistent websocket connection.

> The question is, can the browser-server session be a websocket session so all the communication takes place as a websocket and not http?

Yes. That is indeed the default behaviour when using NiceGUI.

**Reply by dineshbvadhia** [reply]
`So in your scenario the endpoints are implemented with FastAPI "inside" the NiceGUI app? Only one server?`
Yes and yes

Thank-you. Will close as answer.
